File: parsers/parser.php
<?php
/**
 * PHP парсер для витягування запитань з HTML-сторінок
 *
 * Використання:
 *   php parsers/parser.php
 *
 * Результат зберігається у JSON файл у директорії output/
 * Файл можна використати для імпорту через API або адмін-панель
 */

// Завантаження конфігурації
$config = require __DIR__ . '/config.php';

// =============================================
// Клас парсера
// =============================================
class HtmlQuestionParser {
    private array $config;
    private array $questions = [];

    public function __construct(array $config) {
        $this->config = $config;
    }

    /**
     * Завантаження HTML сторінки
     */
    private function fetchHtml(string $url): string {
        // Налаштування контексту для HTTP запиту
        $options = [
            'http' => [
                'method'     => 'GET',
                'header'     => "User-Agent: Mozilla/5.0 (compatible; SchoolParser/1.0)\r\n",
                'timeout'    => 30,
                'follow_location' => 1,
            ],
            'ssl' => [
                'verify_peer'       => true,
                'verify_peer_name'  => true,
            ],
        ];

        $context = stream_context_create($options);
        $html = @file_get_contents($url, false, $context);

        if ($html === false) {
            throw new RuntimeException("Не вдалося завантажити сторінку: {$url}");
        }

        return $html;
    }

    /**
     * Парсинг HTML та витягування запитань
     */
    public function parse(string $html): array {
        // Вимкнення попереджень DOM
        libxml_use_internal_errors(true);

        $dom = new DOMDocument('1.0', 'UTF-8');
        $dom->loadHTML(mb_convert_encoding($html, 'HTML-ENTITIES', 'UTF-8'));

        libxml_clear_errors();

        $xpath = new DOMXPath($dom);
        $selectors = $this->config['selectors'];

        // Пошук контейнерів запитань
        $questionTag   = $selectors['question_container']['tag'];
        $questionClass = $selectors['question_container']['class'];

        // XPath вираз для пошуку контейнерів з заданим класом
        $xpathQuery = "//{$questionTag}[contains(concat(' ', normalize-space(@class), ' '), ' {$questionClass} ')]";
        $questionNodes = $xpath->query($xpathQuery);

        if ($questionNodes === false || $questionNodes->length === 0) {
            echo "Попередження: Контейнери запитань не знайдені. Перевірте конфігурацію.\n";
            return [];
        }

        echo "Знайдено контейнерів запитань: {$questionNodes->length}\n";

        foreach ($questionNodes as $questionNode) {
            $questionData = $this->parseQuestion($xpath, $questionNode);
            if ($questionData !== null) {
                $this->questions[] = $questionData;
            }
        }

        return $this->questions;
    }

    /**
     * Парсинг одного запитання
     */
    private function parseQuestion(DOMXPath $xpath, DOMNode $questionNode): ?array {
        $selectors = $this->config['selectors'];

        // Отримання тексту запитання
        $textTag   = $selectors['question_text']['tag'];
        $textClass = $selectors['question_text']['class'];

        $textQuery = ".//{$textTag}[contains(concat(' ', normalize-space(@class), ' '), ' {$textClass} ')]";
        $textNodes = $xpath->query($textQuery, $questionNode);

        if ($textNodes === false || $textNodes->length === 0) {
            // Якщо клас не знайдено, спробувати просто за тегом
            $textNodes = $xpath->query(".//{$textTag}", $questionNode);
        }

        $questionText = '';
        if ($textNodes !== false && $textNodes->length > 0) {
            $questionText = trim($textNodes->item(0)->textContent);
        }

        if (empty($questionText)) {
            return null;
        }

        // Отримання варіантів відповідей
        $answersTag   = $selectors['answers_list']['tag'];
        $answersClass = $selectors['answers_list']['class'];
        $answerTag    = $selectors['answer_item']['tag'];

        $answersQuery = ".//{$answersTag}[contains(concat(' ', normalize-space(@class), ' '), ' {$answersClass} ')]";
        $answersNodes = $xpath->query($answersQuery, $questionNode);

        if ($answersNodes === false || $answersNodes->length === 0) {
            return null;
        }

        $answersList = $xpath->query(".//{$answerTag}", $answersNodes->item(0));
        if ($answersList === false || $answersList->length === 0) {
            return null;
        }

        $answers = [];
        $correctMarker = $selectors['correct_marker'];

        foreach ($answersList as $answerNode) {
            $answerText = trim($answerNode->textContent);
            if (empty($answerText)) continue;

            // Визначення правильності відповіді
            $isCorrect = false;
            if ($correctMarker['type'] === 'class') {
                $nodeClass = $answerNode->getAttribute('class');
                $isCorrect = str_contains($nodeClass, $correctMarker['value']);
            } elseif ($correctMarker['type'] === 'attribute') {
                $isCorrect = $answerNode->hasAttribute($correctMarker['value']);
            }

            $answers[] = [
                'text'       => $answerText,
                'is_correct' => $isCorrect,
            ];
        }

        if (empty($answers)) {
            return null;
        }

        return [
            'text'    => $questionText,
            'answers' => $answers,
        ];
    }

    /**
     * Збереження результату у JSON файл
     */
    public function saveToJson(array $questions): string {
        $outputDir  = $this->config['output_dir'];
        $outputFile = $this->config['output_file'];

        if (!is_dir($outputDir)) {
            mkdir($outputDir, 0755, true);
        }

        // Формат сумісний з API імпорту
        $output = [
            'topic_id'   => $this->config['topic_id'],
            'class_id'   => $this->config['class_id'],
            'subject_id' => $this->config['subject_id'],
            'questions'  => $questions,
            'parsed_at'  => date('Y-m-d H:i:s'),
            'source_url' => $this->config['url'],
        ];

        $filePath = $outputDir . '/' . $outputFile;
        $json = json_encode($output, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);

        if (file_put_contents($filePath, $json) === false) {
            throw new RuntimeException("Не вдалося зберегти файл: {$filePath}");
        }

        return $filePath;
    }

    /**
     * Запуск парсингу з URL
     */
    public function run(): void {
        $url = $this->config['url'];
        echo "Завантаження сторінки: {$url}\n";

        try {
            $html = $this->fetchHtml($url);
            echo "Сторінку завантажено (" . strlen($html) . " байт).\n";

            $questions = $this->parse($html);
            $count = count($questions);
            echo "Знайдено запитань: {$count}\n";

            if ($count > 0) {
                $filePath = $this->saveToJson($questions);
                echo "Результат збережено: {$filePath}\n";
                echo "Можна використати для імпорту через API.\n";
            } else {
                echo "Запитань не знайдено. Перевірте конфігурацію селекторів.\n";
            }
        } catch (RuntimeException $e) {
            echo "Помилка: " . $e->getMessage() . "\n";
            exit(1);
        }
    }

    /**
     * Парсинг з HTML рядка (для тестування)
     */
    public function parseFromString(string $html): array {
        return $this->parse($html);
    }
}

// =============================================
// Точка входу
// =============================================
$parser = new HtmlQuestionParser($config);
$parser->run();

File: parsers/parser.py
#!/usr/bin/env python3
"""
Python парсер для витягування запитань з HTML-сторінок.

Залежності:
    pip install requests beautifulsoup4

Використання:
    python parsers/parser.py

Результат зберігається у JSON файл у директорії output/
Файл можна використати для імпорту через API або адмін-панель.
"""

import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path

try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("Помилка: Необхідні пакети не встановлено.")
    print("Встановіть їх командою: pip install requests beautifulsoup4")
    sys.exit(1)

# Завантаження конфігурації
# Додавання директорії парсера до шляху
sys.path.insert(0, str(Path(__file__).parent))

try:
    import config as cfg
except ImportError:
    print("Помилка: Файл config.py не знайдено в директорії parsers/")
    sys.exit(1)


class HtmlQuestionParser:
    """Парсер HTML-сторінок з тестовими запитаннями."""

    def __init__(self):
        self.url        = cfg.URL
        self.topic_id   = cfg.TOPIC_ID
        self.class_id   = cfg.CLASS_ID
        self.subject_id = cfg.SUBJECT_ID
        self.selectors  = cfg.SELECTORS
        self.output_dir = Path(__file__).parent / cfg.OUTPUT_DIR
        self.output_file = cfg.OUTPUT_FILE
        self.questions  = []

    def fetch_html(self, url: str) -> str:
        """Завантаження HTML сторінки."""
        print(f"Завантаження сторінки: {url}")
        try:
            response = requests.get(
                url,
                headers=cfg.REQUEST_HEADERS,
                timeout=cfg.REQUEST_TIMEOUT
            )
            response.raise_for_status()
            response.encoding = response.apparent_encoding or 'utf-8'
            print(f"Сторінку завантажено ({len(response.text)} символів).")
            return response.text
        except requests.RequestException as e:
            raise RuntimeError(f"Не вдалося завантажити сторінку: {e}") from e

    def parse(self, html: str) -> list[dict]:
        """Парсинг HTML та витягування запитань."""
        soup = BeautifulSoup(html, 'html.parser')
        selectors = self.selectors

        # Пошук контейнерів запитань
        question_containers = soup.select(selectors['question_container'])

        if not question_containers:
            print("Попередження: Контейнери запитань не знайдені. Перевірте конфігурацію.")
            return []

        print(f"Знайдено контейнерів запитань: {len(question_containers)}")

        questions = []
        for container in question_containers:
            question_data = self._parse_question(container)
            if question_data:
                questions.append(question_data)

        return questions

    def _parse_question(self, container) -> dict | None:
        """Парсинг одного запитання."""
        selectors = self.selectors

        # Текст запитання
        text_element = container.select_one(selectors['question_text'])
        if not text_element:
            # Спроба без класу — просто за тегом
            tag = selectors['question_text'].split('.')[0]
            text_element = container.find(tag)

        if not text_element:
            return None

        question_text = text_element.get_text(strip=True)
        if not question_text:
            return None

        # Список відповідей
        answers_list = container.select_one(selectors['answers_list'])
        if not answers_list:
            return None

        answer_items = answers_list.select(selectors['answer_item'])
        if not answer_items:
            return None

        answers = []
        correct_marker = selectors['correct_marker']

        for item in answer_items:
            answer_text = item.get_text(strip=True)
            if not answer_text:
                continue

            # Визначення правильності відповіді
            is_correct = False
            if correct_marker['type'] == 'class':
                is_correct = correct_marker['value'] in (item.get('class') or [])
            elif correct_marker['type'] == 'attribute':
                is_correct = item.has_attr(correct_marker['value'])

            answers.append({
                'text': answer_text,
                'is_correct': is_correct
            })

        if not answers:
            return None

        return {
            'text': question_text,
            'answers': answers
        }

    def save_to_json(self, questions: list[dict]) -> str:
        """Збереження результату у JSON файл."""
        self.output_dir.mkdir(parents=True, exist_ok=True)

        # Формат сумісний з API імпорту
        output = {
            'topic_id': self.topic_id,
            'class_id': self.class_id,
            'subject_id': self.subject_id,
            'questions': questions,
            'parsed_at': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            'source_url': self.url,
        }

        file_path = self.output_dir / self.output_file
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(output, f, ensure_ascii=False, indent=2)

        return str(file_path)

    def run(self):
        """Запуск парсингу з URL."""
        try:
            html = self.fetch_html(self.url)

            # Затримка перед парсингом
            time.sleep(cfg.REQUEST_DELAY)

            questions = self.parse(html)
            count = len(questions)
            print(f"Знайдено запитань: {count}")

            if count > 0:
                file_path = self.save_to_json(questions)
                print(f"Результат збережено: {file_path}")
                print("Можна використати для імпорту через API.")
            else:
                print("Запитань не знайдено. Перевірте конфігурацію селекторів.")

        except RuntimeError as e:
            print(f"Помилка: {e}")
            sys.exit(1)

    def parse_from_string(self, html: str) -> list[dict]:
        """Парсинг з HTML рядка (для тестування)."""
        return self.parse(html)


# =============================================
# Точка входу
# =============================================
if __name__ == '__main__':
    parser = HtmlQuestionParser()
    parser.run()
